### PR TITLE
DB wrapper

### PR DESCRIPTION
**Pull request by zx80**
It is convenient to mix queries and a connection. The DB class does that.

### CODE REVIEW

**Comment by nackjicholson** [comment]
Thanks for choosing to put your time into this project with me! It's greatly appreciated!

I know I told you in another thread, but for the benefit of others; I'll be out for a weeks so this might sit around for a while before I review it.

At a brief glance, I'm very excited to look at it on more depth. Thanks!

**Comment by zx80** [comment]
> I know I told you in another thread, but for the benefit of others; I'll be out for a few weeks so this might sit around awhile before I review it.

No hurry… :snail: is :ok:

File: aiosql/db.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        if self._db == "sqlite3":
+            import sqlite3 as db
+
+            return db.connect(self._conn_str, **self._conn_options)
+        elif self._db == "psycopg2":
+            import psycopg2 as db  # type: ignore
+
+            return db.connect(self._conn_str, **self._conn_options)
+        else:
+            # note: aiosql currently supports sqlite & postgres
```
[the remark this answers is not in the document]
Comment (zx80) [reply aiosql/db.py:133]: comment removed.

File: aiosql/db.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        log.info(f"creating DB for {db}")
+        # database connection
+        SQLITE = ("sqlite3", "sqlite")
+        POSTGRES = ("pg", "postgres", "postgresql", "psycopg2")
+        if db in SQLITE:
+            self._db = "sqlite3"
+        elif db in POSTGRES:
+            self._db = "psycopg2"
+        else:
+            raise Exception(f"database {db} is supported")
```
[the remark this answers is not in the document]
Comment (zx80) [reply aiosql/db.py:47]: Yep. Exception adjusted.

File: docs/api.md [hunk 3]
```diff
@@ -9,3 +9,8 @@
 
 ::: aiosql.from_str
     :docstring:
+
+## aiosql.DB
+
+::: aiosql.DB
+    :docstring:
```
[the remark this answers is not in the document]
Comment (zx80) [reply docs/api.md:16]: Done.

**Comment by zx80** [comment]
>     1. I'd need to pull it down and really play with it so I feel comfortable owning it.

Ok.

>     2. I want it to also work with database driver adapters so it isn't limited to postgres and sqlite.

Hmmm. The point is that the class is able to create and recreate a connection at will, without the caller having to know which database is used beyond the arguments that were passed and which possibly come from elsewhere (eg, the  environment, or some other configuration external to the user code). The class has to know the database drivers involved because creating a connection is database-driver specific. Now maybe someone added an adapter for something new, and DB cannot know how to create a connection for these.

I extented the connection parameter so that it can be a function, which would be used when the db is not known to DB.

>     3. I think I might document out the DB class methods a little more.

I added some things.

**Comment by nackjicholson** [comment]
@zx80 Thank you again for contributing this. I hope others will find it useful.

I have decided not to pull this into the project though. It doesn't support both normal and asyncio driver types and ultimately I think that makes it non-essential to support. I would encourage you to break this out into your own repo and publish it to pypi (if you don't I might). Let me know if you do and I'll link to it from the README and/or docs.

Cheers!

**Pull request closed without merging**